src/main/kotlin/app/services/impl/HttpKeyService.kt

```diff
@@ -21,44 +23,65 @@ import java.net.URLEncoder
 @Service
 class HttpKeyService(private val httpClientProvider: HttpClientProvider) : KeyService {
 
+    companion object {
+        val logger: Logger = LoggerFactory.getLogger(HttpKeyService::class.toString())
+
+        // Will retry at 1s, 2s, 4s, 8s, 16s then give up (after a total of 31 secs)
+        const val maxAttempts = 5
+        const val initialBackoffMillis = 1000L
+        const val backoffMultiplier = 2.0
+    }
+
+    @Override
+    @Retryable(value = [DataKeyServiceUnavailableException::class],
+        maxAttempts = maxAttempts,
+        backoff = Backoff(delay = initialBackoffMillis, multiplier = backoffMultiplier))
+    @Throws(DataKeyServiceUnavailableException::class, DataKeyDecryptionException::class)
     override fun decryptKey(encryptionKeyId: String, encryptedKey: String): String {
         logger.info("Decrypting encryptedKey: '$encryptedKey', keyEncryptionKeyId: '$encryptionKeyId'.")
         val cacheKey = "$encryptedKey/$encryptionKeyId"
-        return if (decryptedKeyCache.containsKey(cacheKey)) {
-            decryptedKeyCache[cacheKey]!!
-        }
-        else {
-            httpClientProvider.client().use { client ->
-                val url = """$dataKeyServiceUrl/datakey/actions/decrypt?keyId=${URLEncoder.encode(encryptionKeyId, "US-ASCII")}"""
-                logger.info("url: '$url'.")
-                val httpPost = HttpPost(url)
-                httpPost.entity = StringEntity(encryptedKey, ContentType.TEXT_PLAIN)
-                client.execute(httpPost).use { response ->
-                    return when {
-                        response.statusLine.statusCode == 200 -> {
-                            val entity = response.entity
-                            val text = BufferedReader(InputStreamReader(response.entity.content)).use(BufferedReader::readText)
-                            EntityUtils.consume(entity)
-                            val dataKeyResult = Gson().fromJson(text, DataKeyResult::class.java)
-                            decryptedKeyCache[cacheKey] = dataKeyResult.plaintextDataKey
-                            dataKeyResult.plaintextDataKey
+        try {
+            return if (decryptedKeyCache.containsKey(cacheKey)) {
+                decryptedKeyCache[cacheKey]!!
+            }
+            else {
+                httpClientProvider.client().use { client ->
+                    val dksUrl = "$dataKeyServiceUrl/datakey/actions/decrypt?keyId=${URLEncoder.encode(encryptionKeyId, "US-ASCII")}"
+                    logger.info("dataKeyServiceUrl: '$dksUrl'.")
```

Comment:
This log will be a bit chatty for not much benefit?

Reply:
Good point

Reply:
Ok, might be worth downgrading to `.debug` level but if it's not called often then not too bothered either way